File: 6th-sem/OS/lab4/generalized.cpp
#include <iostream>
#include <cstdlib>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;


int greater(int a,int b) {
    return a > b;
}

class Process {
public:
    int id, burst_time, arrival_time, waiting_time, turn_around_time, time_left, rr_priority, priority_value;
    // Process Constructor
    Process(int id, int burst_time, int arrival_time =  0, int priority_value = 0, int waiting_time = 0, int turn_around_time = 0) {
      this -> id = id;
      this -> burst_time = burst_time;
      this -> arrival_time = arrival_time;
      this -> waiting_time = waiting_time;
      this -> turn_around_time = turn_around_time;
      this -> time_left = burst_time;
      this -> rr_priority = 0;
      this -> priority_value = priority_value;
    }
};


struct comparator
{
    inline bool operator() (const Process& lhs, const Process& rhs)
    {
        /*
        for FCFS     --> arrival_time
        for SJF      --> burst_time
        for SRTF     --> time_left
        for Priority --> priority_value
        for RR       --> rr_priority
        */
        return (lhs.priority_value > rhs.priority_value);
    }
};

void insert_to_pq(priority_queue < Process,vector<Process>, comparator >& p, Process _p, int &priority_counter) {
  _p.rr_priority = ++priority_counter;
  p.push(_p);
}

void setQueue(priority_queue < Process,vector<Process>, comparator >& p, vector<Process>& vec,int timer, int &priority_counter) {
    for(auto i: vec) {
      if(i.arrival_time == timer) {
        insert_to_pq(p, i, priority_counter);
      }
    }
}

vector<Process> generalized(vector<Process>& vec, int tq, bool pre_emptive) {

    vector<Process> result;
    int priority_counter = 0;
    int timer = 0;
    priority_queue < Process,vector<Process>, comparator > p;
    
    //  initialize pqueue with processes arrived at t = 0
    setQueue(p, vec, timer, priority_counter);
    
    int process_left = vec.size();

    while(process_left) {

      //  if pqueue is not empty
      if (!p.empty()) {

        Process p1 = p.top();
        p.pop();
        
        //  if non-preemptive, run till completion, else run till minimum of time quantum and time left
        int run_time = pre_emptive? min(tq, p1.time_left) : p1.time_left;
        p1.time_left = p1.time_left - run_time;
        int old_timer = timer;
        timer += run_time;

        //  add the processes that have come while p1 was executed to the pqueue
        for (int i = old_timer + 1; i <= timer; i++)
          setQueue(p, vec, i, priority_counter);

        //  if p1 completed, add it to result
        if(p1.time_left == 0){
          p1.turn_around_time = timer - p1.arrival_time;
          p1.waiting_time = p1.turn_around_time - p1.burst_time;
          process_left--;
          result.push_back(p1);
        }
        //  if p1 not completed, insert it to pqueue again
        else {
          insert_to_pq(p, p1, priority_counter);
        }
      }
      //  if pqueue is empty, time lapse 1 unit and add the newly arrived processes
      else {
        timer++;
        setQueue(p,vec, timer, priority_counter);
      }
    }
    return result;
}

int main() {

  Process p1(1,11,0,2);
  Process p2(2,28,5,0);
  Process p3(3,2,12,3);
  Process p4(4,10,2,1);
  Process p5(5,16,9,4);
  vector<Process> vec;
  vec.push_back(p1);
  vec.push_back(p2);
  vec.push_back(p3);
  vec.push_back(p4);
  vec.push_back(p5);

  int tq = 1;
  
  bool pre_emptive = false;
  vector<Process> result = generalized(vec, tq, pre_emptive);

  for(auto i:result){
    cout << "pid :" << i.id << endl;
    cout << "Waiting Time :" << i.waiting_time << endl;
    cout << "Turn Around Time :" << i.turn_around_time << endl;
    //cout << "Waiting Time :" << i.waiting_time << endl;
  }
  return 0;
}

File: 6th-sem/OS/lab4/generalized.py
#!/usr/bin/python3
'''
Program to simulate different scheduling alogrithms
'''
from queue import PriorityQueue

def insert_to_ready_queue(ready_queue, criteria, process, priority_counter):
	priority_counter += 1
	process['rr_priority'] = priority_counter
	ready_queue.put((process[criteria], process))
	return priority_counter

def set_queue(ready_queue, criteria, processes, current_time, priority_counter):
	for process in processes:
		if process['arrival_time'] == current_time:
			priority_counter = insert_to_ready_queue(ready_queue, criteria, process, priority_counter)
	return priority_counter

def base_scheduler(processes, time_quantum, is_preemptive, criteria):
	result = []
	priority_counter = 0
	timer = 0
	ready_queue = PriorityQueue()

	priority_counter = set_queue(ready_queue, criteria, processes, timer, priority_counter)
	processes_left = len(processes)

	while processes_left:
		if not ready_queue.empty():
			_, process = ready_queue.get()
			runtime = min(time_quantum, process['time_left']) if is_preemptive else process['time_left']
			process['time_left'] -= runtime
			old_timer = timer
			timer += runtime

			for time in range(old_timer + 1, timer + 1):
				priority_counter = set_queue(ready_queue, criteria, processes, time, priority_counter)

			if process['time_left'] == 0:
				process['turn_around_time'] = timer - process['arrival_time']
				process['waiting_time'] = process['turn_around_time'] - process['burst_time']
				processes_left -= 1
				result.append(process)
			else:
				priority_counter = insert_to_ready_queue(ready_queue, criteria, process, priority_counter)
		else:
			timer += 1
			priority_counter = set_queue(ready_queue, criteria, processes, timer, priority_counter)
			
	return result

def fcfs(processes):
	return base_scheduler(processes, 1, False, 'arrival_time')

def sjf(processes):
	return base_scheduler(processes, 1, False, 'burst_time')

def srtf(processes):
	return base_scheduler(processes, 1, True, 'time_left')

def rr(processes):
	return base_scheduler(processes, int(input('Enter time quantum : ')), True, 'rr_priority')

def priority_non_preemptive(processes):
	return base_scheduler(processes, 1, False, 'priority_value')

def priority_preemptive(processes):
	return base_scheduler(processes, 1, True, 'priority_value')

def switcher(scheduling_algo, processes):
	return {
		'1': fcfs,
		'2': sjf,
		'3': srtf,
		'4': rr,
		'5': priority_non_preemptive,
		'6': priority_preemptive
	}[scheduling_algo](processes)

def main():
	processes = []
	for _ in range(int(input("Enter number of processes :- "))):
		print("")
		process = {}
		process['id'] = _
		process['arrival_time'] = int(input("Enter arrival time of p%d:- " % _))
		process['burst_time'] = int(input("Enter burst time of p%d:- " % _))
		process['priority_value'] = int(input("Enter priority value of p%d:- " % _))
		process['time_left'] = process['burst_time']
		process['waiting_time'] = 0
		process['turn_around_time'] = 0
		process['rr_priority'] = 0
		processes.append(process)

	print("Select your scheduling algorithm")
	print("1. FCFS")
	print("2. SJF")
	print("3. SRTF")
	print("4. RR")
	print("5. Priority non-preemptive")
	print("6. Priority preemptive")

	result = switcher(input(), processes)
	result = sorted(result, key=lambda process: process['id'])
	print("PID\tAT\tBT\tPV\tTAT\tWT")
	for process in result:
		print("%d\t%d\t%d\t%d\t%d\t%d\t" % (process['id'], process['arrival_time'], process['burst_time'], process['priority_value'], process['turn_around_time'], process['waiting_time']))

if __name__ == "__main__":
	main()
